### PR TITLE
openocd-rp2040: init at version 0.12.0

### PR DESCRIPTION
**Pull request by luiswirth**
###### Description of changes

- add maintainer lu15w1r7h
- add new package openocd-rp2040 v0.12.0

openocd-rp2040 is a Raspberry Pi fork of openocd which supports the rp2040 chip.
It can be found on GitHub [raspberrypi/openocd](https://github.com/raspberrypi/openocd).

This derivation was made based on the standard openocd derivation.

This is my first ever nixos derivation and my first nix related PR.
Thanks for letting me know if anything is not as it should be and for helping
me figure things out.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/embedded/openocd-rp2040/default.nix [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+, autoconf
+, automake
+, texinfo
+, libtool
+, pkg-config
+, hidapi
+, libftdi1
+, libusb1
+, libgpiod
+}:
```
Comment (MoritzBoehme) [inline pkgs/development/embedded/openocd-rp2040/default.nix:15]:
```suggestion
{ lib
, stdenv
, fetchgit
, pkg-config
, hidapi
, libftdi1
, libusb1
, which
, libtool
, autoconf
, automake
, texinfo
, git
, libgpiod
}:
```
> Arguments should be listed in the order they are used, with the exception of lib, which always goes first.

See [nixpkgs-manual](https://nixos.org/manual/nixpkgs/stable/#sec-syntax)

**Review by MoritzBoehme — CHANGES_REQUESTED** [review]

File: pkgs/development/embedded/openocd-rp2040/default.nix (hunk 1 above)
Comment (LU15W1R7H) [reply pkgs/development/embedded/openocd-rp2040/default.nix]: thanks

File: maintainers/maintainer-list.nix [hunk 2]
```diff
@@ -14725,4 +14725,10 @@
     github = "npatsakula";
     githubId = 23001619;
   };
+  lu15w1r7h = {
+    email = "lwirth2000@gmail.com";
+    name = "Luis Wirth";
+    github = "LU15W1R7H";
+    githubId = 37505890;
+  };
```
Comment (Janik-Haag) [inline maintainers/maintainer-list.nix:14733]: please make this it's own commit.

File: pkgs/development/embedded/openocd-rp2040/default.nix [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+    ./configure --prefix=$out --enable-picoprobe
+  '';
+
+  postInstall = lib.optionalString stdenv.isLinux ''
+    mkdir -p "$out/etc/udev/rules.d"
+    rules="$out/share/openocd/contrib/60-openocd.rules"
+    if [ ! -f "$rules" ]; then
+        echo "$rules is missing, must update the Nix file."
+        exit 1
+    fi
```
Comment (Janik-Haag) [inline pkgs/development/embedded/openocd-rp2040/default.nix:47]: that check seems useless for running the package since there should always be rules files as long as no one updates the package. Maybe add a note in the long description instead telling users to use `services.udev.packages` for this package instead?

File: pkgs/development/embedded/openocd-rp2040/default.nix [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+}:
+
+stdenv.mkDerivation rec {
+  pname = "openocd-rp2040";
+  version = "0.11.0";
+  src = fetchgit {
+    url = "https://github.com/raspberrypi/openocd";
+    rev = "228ede43db3665e470d2e518730de013a8c74411";
+    sha256 = "sha256-NrN01BdkLlM9nHilpOSarJvlDWFCVMVInLP/l7GoCIc=";
+    leaveDotGit = true;
```
Comment (Janik-Haag) [inline pkgs/development/embedded/openocd-rp2040/default.nix:24]: why are you leaving the dot git folder?

**Review by Janik-Haag — CHANGES_REQUESTED** [review]

**Comment by Janik-Haag** [comment]
Do you need help with some of the stuff?

File: pkgs/development/embedded/openocd-rp2040/default.nix (hunk 3 above)
Comment (LU15W1R7H) [reply pkgs/development/embedded/openocd-rp2040/default.nix:47]: I'm doing this to match https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/embedded/openocd/default.nix

They have these lines of code and they also make sense for the rp2040 fork, so I would be inclined to do it the same way as they do. In general I would like to copy their code as much as possible, since the openocd derivation is better maintained than this openocd-rp2040 one.

File: pkgs/development/embedded/openocd-rp2040/default.nix (hunk 4 above)
Comment (LU15W1R7H) [reply pkgs/development/embedded/openocd-rp2040/default.nix:24]: It's been quite a while, since I've written this derivation, but I'm pretty sure that the call to `./bootstrap`  does some git stuff, so it relies on the dotgit folder still being there.

**Comment by LU15W1R7H** [comment]
I can successfully build the derivation with `nix-build -A openocd-rp2040`, but `nixpkgs-review rev head` fails...

```
error: builder for '/nix/store/yfk100p5q966a60yyrjdc9yh8jjasv7k-openocd-8e3c38f.drv' failed with exit code 128;
       last 10 log lines:
       > Submodule 'jimtcl' (https://github.com/msteveb/jimtcl.git) registered for path 'jimtcl'
       > Submodule 'src/jtag/drivers/libjaylink' (https://gitlab.zapb.de/libjaylink/libjaylink.git) registered for path 'src/jtag/drivers/libjaylink'
       > Submodule 'tools/git2cl' (https://git.savannah.nongnu.org/git/git2cl.git) registered for path 'tools/git2cl'
       > Cloning into '/nix/store/rm28gnm0z97sx3sjhwbs2imdl7zzkzki-openocd-8e3c38f/jimtcl'...
       > Cloning into '/nix/store/rm28gnm0z97sx3sjhwbs2imdl7zzkzki-openocd-8e3c38f/src/jtag/drivers/libjaylink'...
       > Cloning into '/nix/store/rm28gnm0z97sx3sjhwbs2imdl7zzkzki-openocd-8e3c38f/tools/git2cl'...
       > Submodule path 'jimtcl': checked out '0aa0fb4e3a38d38a49de9eb585d93d63a370dcf6'
       > Submodule path 'src/jtag/drivers/libjaylink': checked out '9aa7a5957c07bb6e862fc1a6d3153d109c7407e4'
       > Submodule path 'tools/git2cl': checked out '8373c9f74993e218a08819cbcdbab3f3564bbeba'
       > fatal: not a git repository (or any of the parent directories): .git
       For full logs, run 'nix log /nix/store/yfk100p5q966a60yyrjdc9yh8jjasv7k-openocd-8e3c38f.drv'.
error: 1 dependencies of derivation '/nix/store/nn00cr8abfghn9v7skknm7vydsjr0zb5-openocd-rp2040-0.11.0.drv' failed to build
error: 1 dependencies of derivation '/nix/store/0qm5438adh5g0dyppllykn8w09rv86wb-review-shell.drv' failed to build
1 package failed to build:
openocd-rp2040
```

I don't understand the issue. It seems that a `.git` folder is missing...

**Comment by Janik-Haag** [comment]
Seems like the hash upstream changed: 
```
error: hash mismatch in fixed-output derivation '/nix/store/32v4iqlhj5imfq6j2hwhx05mqjaanyy3-openocd-8e3c38f.drv':
         specified: sha256-xNYDDFtplfw1rJmUeHFDxSoUC5F+xUUJWrgv2mIZhkk=
            got:    sha256-4pY4Kzad53mmnnJDLl2FBDuulUldIo1NcNYzZKSCZts=
error: 1 dependencies of derivation '/nix/store/v295cpshj8n4csxr1q8pd22hrdi66myi-openocd-rp2040-0.11.0.drv' failed to build
```

**Comment by Janik-Haag** [comment]
It now builds fine and seems to work as expected.

**Comment by Janik-Haag** [comment]
Result of `nixpkgs-review pr 179713` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>openocd-rp2040</li>
  </ul>
</details>

File: pkgs/development/embedded/openocd-rp2040/default.nix (hunk 4 above)
Comment (Janik-Haag) [reply pkgs/development/embedded/openocd-rp2040/default.nix:24]: We usually want to avoid using `leaveDotGit` as much as possible to the point where we write patches to patch out sections that would need it since `leaveDotGit` isn't really reproducible.

File: pkgs/development/embedded/openocd-rp2040/default.nix (hunk 3 above)
Comment (Janik-Haag) [reply pkgs/development/embedded/openocd-rp2040/default.nix:45]: fair enough

**Comment by LU15W1R7H** [comment]
I'm currently trying out if I can remove the `leaveDotGit`, but I keep running into different issues.

Apparently the git hash has changed again. How can this happen, if the revision stays the same? I'm very confused. I'm now at
`sha256-VtaqIgf63t7/o8KSxqYUgPcwj0crxQSTOCbmepx+vPw=`

After removing `leaveDotGit` the build `nix-build -A openocd-rp2040` failed and I wanted to put it back in. But now the build still fails with the same error as without the `leaveDotGit`
```
> fatal: not a git repository (or any of the parent directories): .git
```
Do I have to clean something such that the build starts from a clean slate?

**Comment by Janik-Haag** [comment]
Sorry that I didn't respond :see_no_evil:

> Do I have to clean something such that the build starts from a clean slate?

Usually not maybe you ran into some borged state, cleaning your local cache might help. You could do that with `nix-collect-garbage --delete-old`

> Apparently the git hash has changed again. How can this happen, if the revision stays the same? I'm very confused. I'm now at
sha256-VtaqIgf63t7/o8KSxqYUgPcwj0crxQSTOCbmepx+vPw=

people force pushing some stuff.

> After removing leaveDotGit the build nix-build -A openocd-rp2040 failed and I wanted to put it back in. But now the build still fails with the same error as without the leaveDotGit

that doesn't happen to me, it works fine locally without leaveDotGit.

File: pkgs/development/embedded/openocd-rp2040/default.nix [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+  src = fetchgit {
+    url = "https://github.com/raspberrypi/openocd";
+    rev = "8e3c38f78730ce878ff81448bc3f6c1e6bb06e13";
+    sha256 = "sha256-4pY4Kzad53mmnnJDLl2FBDuulUldIo1NcNYzZKSCZts=";
+  };
+
+  nativeBuildInputs = [ pkg-config ];
+
+  buildInputs = [ hidapi libftdi1 libusb1 which libtool autoconf automake texinfo git ]
+    ++ lib.optional stdenv.isLinux libgpiod;
```
Comment (Janik-Haag) [inline pkgs/development/embedded/openocd-rp2040/default.nix:29]:
```suggestion
  nativeBuildInputs = [
    pkg-config
  ];

  buildInputs = [
    hidapi
    libftdi1
    libusb1
    which
    libtool
    autoconf
    automake
    texinfo
    git
  ] ++ lib.optional stdenv.isLinux libgpiod;
```

just some formatting nitpicks.

File: pkgs/development/embedded/openocd-rp2040/default.nix [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+}:
+
+stdenv.mkDerivation {
+  pname = "openocd-rp2040";
+  version = "0.11.0";
+  src = fetchgit {
+    url = "https://github.com/raspberrypi/openocd";
+    rev = "8e3c38f78730ce878ff81448bc3f6c1e6bb06e13";
+    sha256 = "sha256-4pY4Kzad53mmnnJDLl2FBDuulUldIo1NcNYzZKSCZts=";
+  };
```
Comment (Janik-Haag) [inline pkgs/development/embedded/openocd-rp2040/default.nix:24]:
```suggestion

  src = fetchFromGitHub {
    owner = "raspberrypi";
    repo = "openocd";
    rev = "8e3c38f78730ce878ff81448bc3f6c1e6bb06e13";
    hash = "sha256-r0UcJ5/a51NLHtrTIBmWHZJ/ucVMB+NoJoTMtBRI9xM=";
  };
```

Comment (Janik-Haag) [reply pkgs/development/embedded/openocd-rp2040/default.nix:24]: actually the hash changed again, seems like upstream does do some whonky stuff, they also don't use tags for releases :confused: maybe open a issue and ask them to do proper releases.

Comment (LU15W1R7H) [reply pkgs/development/embedded/openocd-rp2040/default.nix:24]: I'm trying to get in touch with the maintainers. raspberrypi/openocd#83 Let's see if they know anything and if we can resolve this with them.

File: pkgs/development/embedded/openocd-rp2040/default.nix (hunk 4 above)
Comment (LU15W1R7H) [reply pkgs/development/embedded/openocd-rp2040/default.nix]: I'm unable to remove `leaveDotGit`.
```
Setting up submodules
fatal: not a git repository (or any of the parent directories): .git
error: builder for '/nix/store/jvz4qr7nad375lm3iq5fpgrck0i826vg-openocd-rp2040-0.12.0.drv' failed with exit code 128;
       last 10 log lines:
       > configure.ac:42: installing 'build-aux/config.guess'
       > configure.ac:42: installing 'build-aux/config.sub'
       > configure.ac:19: installing 'build-aux/install-sh'
       > configure.ac:19: installing 'build-aux/missing'
       > Makefile.am: installing './INSTALL'
       > Makefile.am: installing 'build-aux/depcomp'
       > Makefile.am:26: installing 'build-aux/mdate-sh'
       > Makefile.am:26: installing 'build-aux/texinfo.tex'
       > Setting up submodules
       > fatal: not a git repository (or any of the parent directories): .git
```

**Comment by LU15W1R7H** [comment]
I'm not sure how to go forward from this. I'm unable to remove `leaveDotGit` and at the same time, we keep on running into bugs related to it...

**Comment by lurch** [comment]
I don't know how NixOS works, but if I do this I can get it to build whilst avoiding your `fatal: not a git repository` error:
```bash
$ cd $(mktemp -d)
$ git clone https://github.com/raspberrypi/openocd
Cloning into 'openocd'...
remote: Enumerating objects: 77844, done.
remote: Counting objects: 100% (18923/18923), done.
remote: Compressing objects: 100% (3172/3172), done.
remote: Total 77844 (delta 15836), reused 18651 (delta 15697), pack-reused 58921
Receiving objects: 100% (77844/77844), 22.09 MiB | 16.41 MiB/s, done.
Resolving deltas: 100% (63734/63734), done.
$ cd openocd
$ git submodule update --init
Submodule 'jimtcl' (https://github.com/msteveb/jimtcl.git) registered for path 'jimtcl'
Submodule 'src/jtag/drivers/libjaylink' (https://gitlab.zapb.de/libjaylink/libjaylink.git) registered for path 'src/jtag/drivers/libjaylink'
Submodule 'tools/git2cl' (https://git.savannah.nongnu.org/git/git2cl.git) registered for path 'tools/git2cl'
Cloning into '/tmp/tmp.1i4cwB4bA2/openocd/jimtcl'...
Cloning into '/tmp/tmp.1i4cwB4bA2/openocd/src/jtag/drivers/libjaylink'...
Cloning into '/tmp/tmp.1i4cwB4bA2/openocd/tools/git2cl'...
Submodule path 'jimtcl': checked out 'a77ef1a6218fad4c928ddbdc03c1aedc41007e70'
Submodule path 'src/jtag/drivers/libjaylink': checked out '0d23921a05d5d427332a142d154c213d0c306eb1'
Submodule path 'tools/git2cl': checked out '8373c9f74993e218a08819cbcdbab3f3564bbeba'
$ rm -rf .git
$ SKIP_SUBMODULE=1 ./bootstrap 
+ aclocal --warnings=all
+ libtoolize --automake --copy
+ autoconf --warnings=all
+ autoheader --warnings=all
+ automake --warnings=all --gnu --add-missing --copy
configure.ac:24: installing 'build-aux/compile'
configure.ac:42: installing 'build-aux/config.guess'
configure.ac:42: installing 'build-aux/config.sub'
configure.ac:19: installing 'build-aux/install-sh'
configure.ac:19: installing 'build-aux/missing'
Makefile.am: installing './INSTALL'
Makefile.am: installing 'build-aux/depcomp'
Makefile.am:26: installing 'build-aux/mdate-sh'
Makefile.am:26: installing 'build-aux/texinfo.tex'
Skipping submodule setup
Generating build system...
libtoolize: putting auxiliary files in AC_CONFIG_AUX_DIR, 'build-aux'.
libtoolize: copying file 'build-aux/config.guess'
libtoolize: copying file 'build-aux/config.sub'
libtoolize: copying file 'build-aux/install-sh'
libtoolize: copying file 'build-aux/ltmain.sh'
libtoolize: putting macros in AC_CONFIG_MACRO_DIRS, 'm4'.
libtoolize: copying file 'm4/libtool.m4'
libtoolize: copying file 'm4/ltoptions.m4'
libtoolize: copying file 'm4/ltsugar.m4'
libtoolize: copying file 'm4/ltversion.m4'
libtoolize: copying file 'm4/lt~obsolete.m4'
configure.ac:44: installing 'build-aux/ar-lib'
configure.ac:37: installing 'build-aux/compile'
configure.ac:30: installing 'build-aux/missing'
libjaylink/Makefile.am: installing 'build-aux/depcomp'
Bootstrap complete. Quick build instructions:
./configure ....
```

Alternatively, you can do:
```bash
$ cd $(mktemp -d)
$ git clone --recursive https://github.com/raspberrypi/openocd
Cloning into 'openocd'...
remote: Enumerating objects: 77844, done.
remote: Counting objects: 100% (18923/18923), done.
remote: Compressing objects: 100% (3172/3172), done.
remote: Total 77844 (delta 15836), reused 18651 (delta 15697), pack-reused 58921
Receiving objects: 100% (77844/77844), 22.09 MiB | 14.83 MiB/s, done.
Resolving deltas: 100% (63734/63734), done.
Submodule 'jimtcl' (https://github.com/msteveb/jimtcl.git) registered for path 'jimtcl'
Submodule 'src/jtag/drivers/libjaylink' (https://gitlab.zapb.de/libjaylink/libjaylink.git) registered for path 'src/jtag/drivers/libjaylink'
Submodule 'tools/git2cl' (https://git.savannah.nongnu.org/git/git2cl.git) registered for path 'tools/git2cl'
Cloning into '/tmp/tmp.BgtaU6yc6t/openocd/jimtcl'...
remote: Enumerating objects: 9372, done.        
remote: Counting objects: 100% (1157/1157), done.        
remote: Compressing objects: 100% (500/500), done.        
remote: Total 9372 (delta 740), reused 1029 (delta 652), pack-reused 8215        
Receiving objects: 100% (9372/9372), 15.65 MiB | 13.92 MiB/s, done.
Resolving deltas: 100% (6063/6063), done.
Cloning into '/tmp/tmp.BgtaU6yc6t/openocd/src/jtag/drivers/libjaylink'...
remote: Enumerating objects: 1199, done.        
remote: Total 1199 (delta 0), reused 0 (delta 0), pack-reused 1199        
Receiving objects: 100% (1199/1199), 272.88 KiB | 2.84 MiB/s, done.
Resolving deltas: 100% (912/912), done.
Cloning into '/tmp/tmp.BgtaU6yc6t/openocd/tools/git2cl'...
remote: Counting objects: 66, done.        
remote: Compressing objects: 100% (32/32), done.        
remote: Total 66 (delta 34), reused 66 (delta 34)        
Submodule path 'jimtcl': checked out 'a77ef1a6218fad4c928ddbdc03c1aedc41007e70'
Submodule path 'src/jtag/drivers/libjaylink': checked out '0d23921a05d5d427332a142d154c213d0c306eb1'
Submodule path 'tools/git2cl': checked out '8373c9f74993e218a08819cbcdbab3f3564bbeba'
$ cd openocd
$ rm -rf .git
$ SKIP_SUBMODULE=1 ./bootstrap 
+ aclocal --warnings=all
+ libtoolize --automake --copy
+ autoconf --warnings=all
+ autoheader --warnings=all
+ automake --warnings=all --gnu --add-missing --copy
configure.ac:24: installing 'build-aux/compile'
configure.ac:42: installing 'build-aux/config.guess'
configure.ac:42: installing 'build-aux/config.sub'
configure.ac:19: installing 'build-aux/install-sh'
configure.ac:19: installing 'build-aux/missing'
Makefile.am: installing './INSTALL'
Makefile.am: installing 'build-aux/depcomp'
Makefile.am:26: installing 'build-aux/mdate-sh'
Makefile.am:26: installing 'build-aux/texinfo.tex'
Skipping submodule setup
Generating build system...
libtoolize: putting auxiliary files in AC_CONFIG_AUX_DIR, 'build-aux'.
libtoolize: copying file 'build-aux/config.guess'
libtoolize: copying file 'build-aux/config.sub'
libtoolize: copying file 'build-aux/install-sh'
libtoolize: copying file 'build-aux/ltmain.sh'
libtoolize: putting macros in AC_CONFIG_MACRO_DIRS, 'm4'.
libtoolize: copying file 'm4/libtool.m4'
libtoolize: copying file 'm4/ltoptions.m4'
libtoolize: copying file 'm4/ltsugar.m4'
libtoolize: copying file 'm4/ltversion.m4'
libtoolize: copying file 'm4/lt~obsolete.m4'
configure.ac:44: installing 'build-aux/ar-lib'
configure.ac:37: installing 'build-aux/compile'
configure.ac:30: installing 'build-aux/missing'
libjaylink/Makefile.am: installing 'build-aux/depcomp'
Bootstrap complete. Quick build instructions:
./configure ....
```

**Comment by Janik-Haag** [comment]
Okay so the reason why you can only build is when `leaveDotGit` is set to true is because the makefile tries to fetch the submodules based on inforamtion from the .git folder:
```

File: pkgs/development/embedded/openocd-rp2040/default.nix [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+}:
+
+stdenv.mkDerivation {
+  pname = "openocd-rp2040";
+  version = "0.12.0";
+  src = fetchgit {
+    url = "https://github.com/raspberrypi/openocd";
+    rev = "4d87f6dcae77d3cbcd8ac3f7dc887adf46ffa504";
+    sha256 = "sha256-4pY4Kzad53mmnnJDLl2FBDuulUldIo1NcNYzZKSCZts=";
+  };
```
Comment (Janik-Haag) [inline pkgs/development/embedded/openocd-rp2040/default.nix:24]:
```suggestion
  src = fetchgit {
    url = "https://github.com/raspberrypi/openocd";
    rev = "4d87f6dcae77d3cbcd8ac3f7dc887adf46ffa504";
    sha256 = "sha256-4pY4Kzad53mmnnJDLl2FBDuulUldIo1NcNYzZKSCZts=";
    fetchSubmodules = true
  };
```

Comment (Janik-Haag) [reply pkgs/development/embedded/openocd-rp2040/default.nix:24]: For more information look at the fetchGit section: https://nixos.org/manual/nixpkgs/stable/#fetchgit

**Comment by lurch** [comment]
> so we will have to patch the makefile

Yes, or just use `SKIP_SUBMODULE=1 ./bootstrap` instead of `./bootstrap` as I showed above :slightly_smiling_face:

File: pkgs/development/embedded/openocd-rp2040/default.nix (hunk 4 above)
Comment (LU15W1R7H) [reply pkgs/development/embedded/openocd-rp2040/default.nix]: Removed `leaveDotGit`.

**Comment by LU15W1R7H** [comment]
Great thanks! It now works without `leaveDotGit`! I successfully built `openocd-rp2040` and ran it.
However when doing `nixpkgs-review pr 179713`  it fails with
```
/nix/store/ir0j7zqlw9dc49grmwplppc7gh0s40yf-bash-5.2-p15/bin/bash ./libtool  --tag=CC   --mode=compile gcc -DHAVE_CONFIG_H -I.   -I./src -I./src -DPKGDATADIR=\"/nix/store/9rxgqla7jf48vmysi2m1gnsm9i8ihvqa-openo>
src/jtag/drivers/linuxgpiod.c:348:26: error: 'GPIOD_LINE_REQUEST_FLAG_ACTIVE_LOW' undeclared (first use in this function)
  348 |                 flags |= GPIOD_LINE_REQUEST_FLAG_ACTIVE_LOW;
      |                          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
src/jtag/drivers/linuxgpiod.c:350:16: error: variable 'config' has initializer but incomplete type
  350 |         struct gpiod_line_request_config config = {
      |                ^~~~~~~~~~~~~~~~~~~~~~~~~
src/jtag/drivers/linuxgpiod.c:351:18: error: 'struct gpiod_line_request_config' has no member named 'consumer'
  351 |                 .consumer = "OpenOCD",
      |                  ^~~~~~~~
src/jtag/drivers/linuxgpiod.c:351:29: error: excess elements in struct initializer [-Werror]
  351 |                 .consumer = "OpenOCD",
      |                             ^~~~~~~~~
src/jtag/drivers/linuxgpiod.c:351:29: note: (near initialization for 'config')
src/jtag/drivers/linuxgpiod.c:352:18: error: 'struct gpiod_line_request_config' has no member named 'request_type'
  352 |                 .request_type = dir,
      |                  ^~~~~~~~~~~~
src/jtag/drivers/linuxgpiod.c:352:33: error: excess elements in struct initializer [-Werror]
  352 |                 .request_type = dir,
      |                                 ^~~
src/jtag/drivers/linuxgpiod.c:352:33: note: (near initialization for 'config')
src/jtag/drivers/linuxgpiod.c:353:18: error: 'struct gpiod_line_request_config' has no member named 'flags'
  353 |                 .flags = flags,
      |                  ^~~~~
src/jtag/drivers/linuxgpiod.c:353:26: error: excess elements in struct initializer [-Werror]
  353 |                 .flags = flags,
      |                          ^~~~~
src/jtag/drivers/linuxgpiod.c:353:26: note: (near initialization for 'config')
src/jtag/drivers/linuxgpiod.c:350:42: error: storage size of 'config' isn't known
  350 |         struct gpiod_line_request_config config = {
      |                                          ^~~~~~
src/jtag/drivers/linuxgpiod.c:356:18: error: implicit declaration of function 'gpiod_line_request'; did you mean 'gpiod_line_request_get_fd'? [-Werror=implicit-function-declaration]
  356 |         retval = gpiod_line_request(gpiod_line[idx], &config, val);
      |                  ^~~~~~~~~~~~~~~~~~
      |                  gpiod_line_request_get_fd
src/jtag/drivers/linuxgpiod.c:350:42: error: unused variable 'config' [-Werror=unused-variable]
  350 |         struct gpiod_line_request_config config = {
      |                                          ^~~~~~
mv -f src/jtag/drivers/.deps/libocdjtagdrivers_la-libusb_helper.Tpo src/jtag/drivers/.deps/libocdjtagdrivers_la-libusb_helper.Plo
cc1: all warnings being treated as errors
```

**Comment by LU15W1R7H** [comment]
@bjornfor That was it! `nixpkgs-review` is now happy!

**Review by Janik-Haag — COMMENTED** [review]
This should be good to go once you squash the [openocd-rp2040: fixup pin libgpiod](https://github.com/NixOS/nixpkgs/pull/179713/commits/5866fa2f885757fdfda12e36a4c7b89d39af9290) in to the init commit since nixpkgs tries to have  atomic commit history.

**Comment by LU15W1R7H** [comment]
Done!

**Review by Janik-Haag — APPROVED** [review]

**Comment by Janik-Haag** [comment]
Darwin builds are failing, please either disabling them by setting `platforms = platforms.linux;` in the meta section or `broken = stdenv.isDarwin;` if it's supposed to work but doesn't. Might also be not too hard to fix since the current error message in the build log says it can't find a program.

**Comment by LU15W1R7H** [comment]
I've set `platforms = platforms.linux`, because I don't have a Mac and can't say anything about it's support.

File: pkgs/development/embedded/openocd-rp2040/default.nix [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+  meta = with lib; {
+    description = "OpenOCD fork for rp2040 microcontroller";
+    longDescription = ''
+      This is a fork of OpenOCD by Raspberry Pi,
+      which brings support to the rp2040 microcontroller.
+    '';
+    homepage = "https://github.com/raspberrypi/openocd";
+    license = licenses.gpl2Plus;
+    maintainers = with maintainers; [ lu15w1r7h ];
+    platforms = platforms.unix;
```
Comment (Janik-Haag) [inline pkgs/development/embedded/openocd-rp2040/default.nix:79]: Darwin is also a unix platform so you will have to limit it to 
```suggestion
    platforms = platforms.linux;
```

should be good to go after that.

Comment (LU15W1R7H) [reply pkgs/development/embedded/openocd-rp2040/default.nix]: oops forgot to push the change.

**Comment by Janik-Haag** [comment]
Result of `nixpkgs-review pr 179713` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>openocd-rp2040</li>
  </ul>
</details>

**Comment by Janik-Haag** [comment]
Sorry that it took so long but it seems good to me now.

**Comment by LU15W1R7H** [comment]
nice 🥳

**Comment by Janik-Haag** [comment]
Feel free to ping me if you do another pr, so we maybe can get it merged faster :D
